src/Streaming/Internal/Many.hs

```diff
@@ -98,6 +108,163 @@ unzip = loop
 {-# INLINABLE unzip #-}
 
 
+{- Remarks on the design of zip functions
+
+Zip functions have two design choices:
+(1) What do we do with the end-of-stream values of both streams?
+(2) If the streams are of different length, do we keep or throw out the
+remainder of the longer stream?
+
+* We are assuming not to take infinite streams as input and instead deal with
+reasonably small finite streams.
+* To avoid making choices for the user, we keep both end-of-stream payloads
+* The default zips (ones without a prime in the name) use @effects@ to consume
+the remainder stream after zipping. We include zip function variants that
+return no remainder (for equal length streams), or the remainder of the
+longer stream.
+
+-}
+
+-- | The remainder of zipping two streams
+type ZipResidual a b m r1 r2 =
+  ( Either r1 (Stream (Of a) m r1)
+  , Either r2 (Stream (Of b) m r2)
+  )
```

Comment:
A more faithful type would be
```suggestion
type ZipResidual a b m r1 r2 = Either
  ((Stream (Of a) m r1), r2)
  (r1, (Stream (Of b) m r2))
```

Reply:
This is the wrong type. You could have `(r1, r2)` as the remainder if the streams are equal length. My type has the issue of allowing two remainders. Your type has the issue of omitting the no-remainder case. We'd need a type that says you can have none, or one remainder but not both. I'll write that instead. I agree and it's way better to have tighter types.

Reply:
I replaced it with `Either3` but a simple residual type for `zipWith3R` isn't easy (and `Either7` isn't pretty) so I'm leaving it overly liberal for now rather than too specific (i.e., implementing your residual type would omit a valid case).

Also, the original implementation didn't check for the `(r1,r2)` and was thus incorrect :disappointed:  -- leading to this discussion.